ale_linters/cpp/clang.vim

```diff
@@ -9,15 +10,31 @@ function! ale_linters#cpp#clang#GetExecutable(buffer) abort
 endfunction
 
 function! ale_linters#cpp#clang#GetCommand(buffer) abort
-    let l:paths = ale#c#FindLocalHeaderPaths(a:buffer)
-
-    " -iquote with the directory the file is in makes #include work for
-    "  headers in the same directory.
-    return ale#Escape(ale_linters#cpp#clang#GetExecutable(a:buffer))
-    \   . ' -S -x c++ -fsyntax-only '
-    \   . '-iquote ' . ale#Escape(fnamemodify(bufname(a:buffer), ':p:h')) . ' '
-    \   . ale#c#IncludeOptions(l:paths)
-    \   . ale#Var(a:buffer, 'cpp_clang_options') . ' -'
+    let l:build_dir = ale#c#FindCompileCommands(a:buffer)
+    if empty(l:build_dir)
+        let l:paths = ale#c#FindLocalHeaderPaths(a:buffer)
+        " -iquote with the directory the file is in makes #include work for
+        "  headers in the same directory.
+        return ale#Escape(ale_linters#cpp#clang#GetExecutable(a:buffer))
+        \   . ' -S -x c++ -fsyntax-only '
+        \   . '-iquote ' . ale#Escape(fnamemodify(bufname(a:buffer), ':p:h')) . ' '
+        \   . ale#c#IncludeOptions(l:paths)
+        \   . ale#Var(a:buffer, 'cpp_clang_options') . ' -'
+    else
+        let l:c = readfile(l:build_dir . '/compile_commands.json')
+        let l:l = json_decode(l:c)
+        let l:b = fnamemodify(bufname(a:buffer), ':p')
+        for l:d in l:l
+            if l:d["file"] == l:b
```

Comment:
The filenames in the compile_commands.json files generated by ninja (and therefore meson) are relative paths, so l:d["file"] should be expanded to an absolute path as well to compare.